src/Compilers/CSharp/Test/Symbol/Symbols/Source/RecordTests.cs

```diff
@@ -588,6 +609,90 @@ .maxstack  3
   IL_0049:  ret
   IL_004a:  ldc.i4.0
   IL_004b:  ret
+}");
+        }
+
+        [Fact]
+        public void RecordClone1()
+        {
+            var comp = CreateCompilation("data class C(int x, int y);");
+            comp.VerifyDiagnostics();
+
+            var c = comp.GlobalNamespace.GetTypeMember("C");
+            var clone = c.GetMethod(WellKnownMemberNames.CloneMethodName);
+            Assert.Equal(0, clone.Arity);
+            Assert.Equal(0, clone.ParameterCount);
+            Assert.Equal(c, clone.ReturnType);
+
+            var ctor = (MethodSymbol)c.GetMembers(".ctor")[1];
+            Assert.Equal(1, ctor.ParameterCount);
+            Assert.True(ctor.Parameters[0].Type.Equals(c, TypeCompareKind.ConsiderEverything));
+
+            var verifier = CompileAndVerify(comp, verify: Verification.Fails);
+            verifier.VerifyIL("C." + WellKnownMemberNames.CloneMethodName, @"
+{
+  // Code size        7 (0x7)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  newobj     ""C..ctor(C)""
+  IL_0006:  ret
+}
+");
+            verifier.VerifyIL("C..ctor(C)", @"
+{
+  // Code size       31 (0x1f)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  call       ""object..ctor()""
+  IL_0006:  ldarg.0
+  IL_0007:  ldarg.1
+  IL_0008:  ldfld      ""int C.<x>k__BackingField""
+  IL_000d:  stfld      ""int C.<x>k__BackingField""
+  IL_0012:  ldarg.0
+  IL_0013:  ldarg.1
+  IL_0014:  ldfld      ""int C.<y>k__BackingField""
+  IL_0019:  stfld      ""int C.<y>k__BackingField""
+  IL_001e:  ret
+}");
+        }
+
+        [Fact]
+        public void RecordClone2()
+        {
+            var comp = CreateCompilation(@"
+data class C(int x, int y)
+{
+    public C(C other) { }
+}");
+            comp.VerifyDiagnostics();
+
+            var c = comp.GlobalNamespace.GetTypeMember("C");
+            var clone = c.GetMethod(WellKnownMemberNames.CloneMethodName);
+            Assert.Equal(0, clone.Arity);
+            Assert.Equal(0, clone.ParameterCount);
+            Assert.Equal(c, clone.ReturnType);
+
+            var ctor = (MethodSymbol)c.GetMembers(".ctor")[0];
+            Assert.Equal(1, ctor.ParameterCount);
+            Assert.True(ctor.Parameters[0].Type.Equals(c, TypeCompareKind.ConsiderEverything));
+
+            var verifier = CompileAndVerify(comp, verify: Verification.Fails);
+            verifier.VerifyIL("C." + WellKnownMemberNames.CloneMethodName, @"
+{
+  // Code size        7 (0x7)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  newobj     ""C..ctor(C)""
+  IL_0006:  ret
+}
+");
+            verifier.VerifyIL("C..ctor(C)", @"
+{
+  // Code size        7 (0x7)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  call       ""object..ctor()""
+  IL_0006:  ret
 }");
         }
```

Comment:
Consider the following tests, or add them to test plan:
1. user provided `Clone()` method with different return type or void or `ref`, then verify members on record type
2. verifying execution order for `with` initializer
3. out vars in `with` initializer (verifies scoping and execution order)
4. `with` used with extension `Clone()` method
5. `with` used with `Clone(params string[] args)`
6. `with` used with `Clone(string optional = null)`
7. static Clone `method()` in code
8. two copy constructors, one is correct and the other adds `params` or optional parameter
9. test with missing `IsExternalInit` type
10. record with user-provided property (get, init, get+init, set, get+set) corresponding to a positional member
11. `with` on type with inherited `Clone` (accessible or private)

Reply:
Ugh, I had a big comment here that Github ate in a refresh.

Short answer: most of the Clone() tests are less necessary because Clone will be unspeakable according to LDM. The tests will only be useful as a test of metadata resilience.

(12), (10), (9), (2), (7), and (10) all have tests.

Reply:
I'll add the remaining two to test plan.

3. out vars in with initializer (verifies scoping and execution order)
8. two copy constructors, one is correct and the other adds params or optional parameter

I'm surprised that we won't let users customize Clone method. Discussion for another day ;-)

---
In reply to: [432127584](https://github.com/dotnet/roslyn/pull/44582#discussion_r432127584) [](ancestors = 432127584)